src/pudl/extract/dbf.py

```diff
@@ -70,6 +72,108 @@ def create_sa_table(self, sa_meta: sa.MetaData) -> sa.Table:
+            dbf = DBF(
+                "",
+                ignore_missing_memofile=True,
+                filedata=self.zipfile.open(self.dbc_path.as_posix()),
+            )
+            table_names: dict[Any, str] = {}
+            table_columns = defaultdict(list)
+            for row in dbf:
+                obj_id = row.get("OBJECTID")
+                obj_name = row.get("OBJECTNAME")
+                obj_type = row.get("OBJECTTYPE", None)
+                if obj_type == "Table":
+                    table_names[obj_id] = obj_name
+                elif obj_type == "Field":
+                    parent_id = row.get("PARENTID")
+                    table_columns[parent_id].append(obj_name)
+            # Remap table ids to table names.
+            self._table_schemas = {
+                table_names[tid]: cols for tid, cols in table_columns.items()
+            }
+        return self._table_schemas
+
+    def get_table_dbf(self, table_name: str) -> DBF:
+        """Opens the DBF for a given table."""
+        fname = self._table_file_map[table_name]
+        return DBF(
+            fname,
+            encoding="latin1",
+            parserclass=self.field_parser,
+            ignore_missing_memofile=True,
+            filedata=self.get_file(fname),
+        )
+
+    @lru_cache
+    def get_table_schema(self, table_name: str) -> DbfTableSchema:
+        """Returns TableSchema for a given table and a given year."""
+        table_columns = self.get_db_schema()[table_name]
+        dbf = self.get_table_dbf(table_name)
+        dbf_fields = [field for field in dbf.fields if field.name != "_NullFlags"]
+        if len(table_columns) != len(table_columns):
+            return ValueError(
+                f"Number of DBF fields in {table_name} does not match what was "
+                f"found in the DBC index file for {self.partition}."
+            )
+        schema = DbfTableSchema(table_name)
+        for long_name, dbf_col in zip(table_columns, dbf_fields):
+            if long_name[:8] != dbf_col.name.lower()[:8]:
+                raise ValueError(
+                    f"DBF field name mismatch: {dbf_col.name} != {long_name}"
+                )
+            col_type = DBF_TYPES[dbf_col.type]
+            if col_type == sa.String:
+                col_type = sa.String(length=dbf_col.length)
+            schema.add_column(long_name, col_type, short_name=dbf_col.name)
+        return schema
+
+    def load_table(self, table_name: str) -> pd.DataFrame:
+        """Returns dataframe that holds data for a table contained within this archive.
+
+        Args:
+            table_name: name of the table.
+        """
+        sch = self.get_table_schema(table_name)
+        df = pd.DataFrame(iter(self.get_table_dbf(table_name)))
+        df = df.drop("_NullFlags", axis=1, errors="ignore").rename(
+            sch.get_column_rename_map(), axis=1
+        )
+        return df
+
+
 class AbstractFercDbfReader(Protocol):
```

Comment:
I think the idea here is that whenever something has these methods, it will be of type `AbstractFercDbfReader` even if it's not explicitly its subclass. In essence, I think this allows us to create "interface definitions" that can then be passed as types to methods/objects that consume this and it can ensure that it's called in the right way.